### PR TITLE
Combine storage: Add publishers for StorageReference

### PR DESCRIPTION
**Pull request by ziadtamim**
This PR adds Combine support for Firebase Storage.

```swift 
  ref?.putFilePublisher(from: dummyFileURL, metadata: nil)
    .sink { completion in
      switch completion {
      case .finished:
        print("Finished")
      case let .failure(error):
        print("\(error.localizedDescription)")
      }
    } receiveValue: { metadata in
       print("\(metadata)")   
    }
    .store(in: &cancellables)
```

```swift 
  ref?.putDataPublisher(from: data, metadata: nil)
    .sink { completion in
      switch completion {
      case .finished:
        print("Finished")
      case let .failure(error):
        print("\(error.localizedDescription)")
      }
    } receiveValue: { metadata in
       print("\(metadata)")   
    }
    .store(in: &cancellables)
```

### CODE REVIEW

**Comment by ziadtamim** [comment]
@googlebot I consent.

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    /// - Parameters:
+    ///   - uploadData: The Data to upload.
+    ///   - metadata: metadata `StorageMetadata` containing additional information (MIME type, etc.)
+    ///
+    /// - Returns: A publisher emitting a `StorageMetadata` instance. The publisher will emit on the *main* thread.
+    @discardableResult
+    public func putDataPublisher(_ data: Data,
+                                 metadata: StorageMetadata?)
+      -> AnyPublisher<StorageMetadata, Error> {
+      let subject = PassthroughSubject<StorageMetadata, Error>()
```
Comment (peterfriese) [inline FirebaseCombineSwift/Sources/Storage/StorageReference.swift:39]: Shouldn't this be a Future instead? We call once, and receive either an error or a metadata object.

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    /// - Parameters:
+    ///   - fileURL: A `URL` representing the system file path of the object to be uploaded.
+    ///   - metadata: `StorageMetadata` containing additional information (MIME type, etc.) about the object being uploaded.
+    ///
+    /// - Returns: A publisher emitting a `StorageMetadata` instance. The publisher will emit on the *main* thread.
+    @discardableResult
+    public func putFilePublisher(from fileURL: URL,
+                                 metadata: StorageMetadata?)
+      -> AnyPublisher<StorageMetadata, Error> {
+      let subject = PassthroughSubject<StorageMetadata, Error>()
```
Comment (peterfriese) [inline FirebaseCombineSwift/Sources/Storage/StorageReference.swift:67]: Same here.

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    /// The publisher will emit events on the **main** thread.
+    ///
+    /// - Parameters:
+    ///   - size: The maximum size in bytes to download. If the download exceeds this size
+    ///     the task will be cancelled and an error will be returned.
+    ///
+    /// - Returns: A publisher emitting a `Data` instance. The publisher will emit on the *main* thread.
+    @discardableResult
+    public func getData(maxSize size: Int64) -> AnyPublisher<Data, Error> {
+      let subject = PassthroughSubject<Data, Error>()
```
Comment (peterfriese) [inline FirebaseCombineSwift/Sources/Storage/StorageReference.swift:97]: Same.

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    /// The publisher will emit events on the **main** thread.
+    ///
+    /// - Parameters:
+    ///   - fileURL: A file system URL representing the path the object should be downloaded to.
+    ///
+    /// - Returns: A publisher emitting a `URL`  pointing to the file path of the downloaded file
+    ///   on success. The publisher will emit on the *main* thread.
+    @discardableResult
+    public func write(toFile fileURL: URL) -> AnyPublisher<URL, Error> {
+      let subject = PassthroughSubject<URL, Error>()
```
Comment (peterfriese) [inline FirebaseCombineSwift/Sources/Storage/StorageReference.swift:123]: Same.

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+// limitations under the License.
+
+#if canImport(Combine) && swift(>=5.0) && canImport(FirebaseStorage)
+
+  import Combine
+  import FirebaseStorage
+
+  @available(swift 5.0)
+  @available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+  extension StorageReference {
```
Comment (peterfriese) [inline FirebaseCombineSwift/Sources/Storage/StorageReference.swift:22]: Some of the methods in this file are named inconsistently. Only publishers should end in "...Publisher".

File: FirebaseCombineSwift/Tests/Unit/Storage/StorageReferenceTests.swift [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  var storage: Storage?
+
+  override func setUp() {
+    let app = FirebaseApp.appForStorageUnitTestsWithName(name: "App")
+    storage = Storage.storage(app: app)
+  }
+
+  func testReferenceWithNonExistentFileFails() {
```
Comment (peterfriese) [inline FirebaseCombineSwift/Tests/Unit/Storage/StorageReferenceTests.swift:42]: I think we need more tests :-)

**Review by peterfriese — CHANGES_REQUESTED** [review]

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift (hunk 1 above)
Comment (ziadtamim) [reply FirebaseCombineSwift/Sources/Storage/StorageReference.swift:39]: @peterfriese I've used `PassthroughSubject` to manage the returned task. I couldn't find any other alternatives to wire the cancellation of a `Future` asynchronous code.

File: FirebaseCombineSwift/Tests/Unit/Storage/StorageReferenceTests.swift (hunk 6 above)
Comment (ziadtamim) [reply FirebaseCombineSwift/Tests/Unit/Storage/StorageReferenceTests.swift:42]: @peterfriese I am waiting for a confirmation from [this PR](https://github.com/firebase/firebase-ios-sdk/pull/7734) to have more tests

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift (hunk 1 above)
Comment (peterfriese) [reply FirebaseCombineSwift/Sources/Storage/StorageReference.swift:39]: I've sent a PR (https://github.com/ziadtamim/firebase-ios-sdk/pull/1) to your branch with a suggestion how to us Futures for this. The test passes.

**Review by djj1109 — APPROVED** [review]

File: FirebaseCombineSwift/Sources/Storage/StorageReference.swift (hunk 1 above)
Comment (ziadtamim) [reply FirebaseCombineSwift/Sources/Storage/StorageReference.swift]: Thanks @peterfriese

**Review by peterfriese — APPROVED** [review]
LGTM

**Pull request merged**